### PR TITLE
Test cases for #21

### PR DESCRIPTION
**Pull request by MajaRet**
Adds unit tests for the `CompletePatternPass`.

### CODE REVIEW

File: src/test/FreeC/Pass/CompletePatternPassTests.hs [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+          input <- expectParseTestFuncDecl
+            "g x = case (f x) :: Foobar of {Foo -> Foo ; Bar -> Bar}"
+          shouldSucceed $ do
+            _ <- defineTestTypeCon "Foobar" 0 ["Foo", "Bar"]
+            _ <- defineTestCon "Foo" 0 "Foobar"
+            _ <- defineTestCon "Bar" 0 "Foobar"
+            _ <- defineTestFunc "f" 1 "Foobar -> Foobar"
+            _ <- defineTestVar "x"
+            checkPatternFuncDecl input
+    it "fails if the case expression's scrutinee is a lambda" $ do
```
Comment (maclement) [inline src/test/FreeC/Pass/CompletePatternPassTests.hs]:
```suggestion
    it "fails if the case expression's scrutinee is a lambda expression" $ do
```

File: src/test/FreeC/Pass/CompletePatternPassTests.hs [hunk 2]
```diff
@@ -0,0 +1,172 @@
+-- name subject to change
+
```
Comment (maclement) [inline src/test/FreeC/Pass/CompletePatternPassTests.hs]:
```suggestion

```

**Review by maclement — CHANGES_REQUESTED** [review]
I only added two very minor suggestions. Besides that i am really happy with the test cases.

Maybe we should add one or two test cases that check if type synonyms are handled correctly.
Some were mentioned in #42.

File: src/test/FreeC/Pass/CompletePatternPassTests.hs [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+import           FreeC.Monad.Class.Testable
+import           FreeC.Pass.CompletePatternPass
+import           FreeC.Test.Environment
+import           FreeC.Test.Parser
+-------------------------------------------------------------------------------
+-- Tests                                                                     --
+-------------------------------------------------------------------------------
+
+testCompletePatternPass :: SpecWith ()
+testCompletePatternPass = describe "FreeC.Pass.CompletePatternPass" $ do
```
Comment (just95) [inline src/test/FreeC/Pass/CompletePatternPassTests.hs]: You don't really need the heading if there is just a single function in your module. However, there should be documentation for the function itself.

```suggestion

-- | Test group for "FreeC.Pass.CompletePatternPass" tests.
testCompletePatternPass :: SpecWith ()
testCompletePatternPass = describe "FreeC.Pass.CompletePatternPass" $ do
```

File: src/test/FreeC/Pass/CompletePatternPassTests.hs [hunk 4]
```diff
@@ -0,0 +1,172 @@
+-- name subject to change
+
+-- | This module contains tests for "FreeC.Pass.CompletePatternPassTests".
```
Comment (just95) [inline src/test/FreeC/Pass/CompletePatternPassTests.hs]:
```suggestion
-- | This module contains tests for "FreeC.Pass.CompletePatternPass".
```

File: src/test/FreeC/PipelineTests.hs [hunk 5]
```diff
@@ -20,3 +21,4 @@ testPipeline = do
   testPartialityAnalysisPass
   testResolverPass
   testTypeInferencePass
+  testCompletePatternPass
```
Comment (just95) [inline src/test/FreeC/PipelineTests.hs]: Try to keep this list in alphabetical order.

File: src/test/FreeC/Pass/CompletePatternPassTests.hs [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+        _ <- defineTestCon "Foo" 0 "Foobar"
+        _ <- defineTestCon "Bar" 0 "Foobar"
+        _ <- defineTestVar "x"
+        checkPatternFuncDecl input
+    it
+        "fails for a faulty case expression used as another case expression's scrutinee"
+      $ do
+          input <-
+            expectParseTestFuncDecl
+              "f x = case ((case x :: Foobar of {Foo -> x} ) :: Foobar) of {Foo -> Foo ; Bar -> Bar}"
```
Comment (just95) [inline src/test/FreeC/Pass/CompletePatternPassTests.hs]: These lines look like they exceed the 80 character limit. Use string concatenation to break up the lines. Also check other long lines in this file.

```suggestion
    it
        (  "fails for a faulty case expression used as another case"
        ++ "expression's scrutinee"
        )
      $ do
          input <-
            expectParseTestFuncDecl
            $  "f x = case ((case x :: Foobar of {Foo -> x} ) :: Foobar) of {"
            ++ "    Foo -> Foo ;"
            ++ "    Bar -> Bar"
            ++ "  }"
```

File: src/test/FreeC/Pass/CompletePatternPassTests.hs [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+      input <-
+        expectParseTestFuncDecl
+          "f x = case x :: Foobar of {Foo -> case x :: Foobar of {Foo -> Foo ; Bar -> Bar} ; Bar -> Bar}"
+      shouldSucceed $ do
+        _ <- defineTestTypeCon "Foobar" 0 ["Foo", "Bar"]
+        _ <- defineTestCon "Foo" 0 "Foobar"
+        _ <- defineTestCon "Bar" 0 "Foobar"
+        _ <- defineTestVar "x"
+        checkPatternFuncDecl input
+
```
Comment (just95) [inline src/test/FreeC/Pass/CompletePatternPassTests.hs]:
```suggestion
```

**Review by just95 — CHANGES_REQUESTED** [review]

**Review by just95 — APPROVED** [review]

**Review by maclement — APPROVED** [review]

**Pull request merged**